Tools/clinic/clinic.py

```diff
@@ -2653,7 +2643,7 @@ class Function:
     name: str
     module: Module | Clinic
     cls: Class | None
-    c_basename: str | None
+    c_basename: str
```

Comment:
In situations where we previously set `c_basename` to `None`, we now set `c_basename` to an empty string.

Does the concept of having an empty-string `c_basename` make sense? Or is it more the case that some functions don't have a `c_basename` at all?

If it's more the latter than the former, we should stick with what the code currently does; the current annotation is more expressive of the fact that there are really "two states" a function can be in: either it has a `c_basename` (in which case the attribute is a `str`), or it doesn't (in which case the attribute is `None`)

Reply:
> In situations where we previously set `c_basename` to `None`, we now set `c_basename` to an empty string.

No, it should be a valid basename with this PR; if it is not, it is a bug :)

Reply:
The point is that we always need a valid C basename; that's currently ensured in the renderer. I moved the code from the renderer into `parse_names()`, so we can ensure a valid C basename right from the start.

Reply:
Yep, I get it now :)